### ISSUE TITLE
Recommended way of exposing C++ data and interfaces to .NET core

### ISSUE THREAD

**Issue by southpolenator** [dotnet/corefx#24444]
Since there is no C++/CLI, what is the recommended way of exposing C++ data and interfaces to .NET core? Is it C style functions and P/Invoke (for cross platform apps, how would you reference different dynamic library names)? Some kind of (inter-process) communication?

**Comment by karelz** [comment]
PInvoke and COM interop are the recommended ways.
Did you consider unifying the dynamic library names across all platforms you need to support? An alternative would be to cross-compile for specific distros ...

**Comment by karelz** [comment]
cc @janvorli

**Comment by southpolenator** [comment]
Is there an example on how to create C++ COM interface for non-Windows OS?
What I meant for different library names is .dll on Windows, .so on Linux, etc...

**Comment by mellinoe** [comment]
> Is it C style functions and P/Invoke

Yes, this is the way you should do it. This will be the easiest and most portable option.

> (for cross platform apps, how would you reference different dynamic library names)?

If you control it, you should ensure that the library names (barring file extension) are identical. If they are not identical, then the problem becomes significantly harder (see https://github.com/dotnet/coreclr/issues/930).

CoreCLR will automatically prepend "lib" to the dll name when resolving PInvokes, and will also append ".so" or ".dylib" depending on what is appropriate for the platform. This means you can write this:

`[DllImport("mylib") void DoThing()`

And CoreCLR will load that function from these names:
* mylib.dll
* mylib.so
* mylib.dylib
* libmylib.so
* libmylib.dylib
You need to ensure that "mylib" is the common name on all platforms, though.

**Comment by danmosemsft** [comment]
If you have to do a lot of C# / C++ interop we recommend that you autogenerate the interop code. There are number of tools that do that, e.g. https://github.com/mono/CppSharp is one built by the Xamarin folks which we recommend.

**Comment by jkotas** [comment]
cc @russellhadley 

>  https://github.com/mono/CppSharp is one built by the Xamarin folks which we recommend.

It would be nice to have an example that shows how to use this tool with .NET Core.
